File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        li=[]
        def bc(s,o,c):
            if(len(s)==n*2):
                li.append(s)
                return
            if(o<n):
                bc(s+'(',o+1,c)
            if(c<o):
                bc(s+')',o,c+1)
        bc("",0,0)
        return li

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    List<String>li=new ArrayList<>();
    public List<String> generateParenthesis(int n) {
        bc("",0,0,n);
        return li;
    }
    public void bc(String s,int o,int c,int n)
    {

        if(s.length()==n*2)
        {
            li.add(s);
        }
        if(o<n)
        {
            bc(s+"(",o+1,c,n);
        }
        if(c<o)
        {
            bc(s+")",o,c+1,n);
        }
    }
}